### PR TITLE
Add CoreConfig + CardClient + approve card order request

### PR DESCRIPTION
**Pull request by minhthenguyen**
### Reason for changes

This PR adds `CoreConfig`, `CardClient` and a few models needed for card payment (`Card`, `CardExpiry`, etc.) to allow merchants to approve a card order:

```
// 1. Init config

let config = CoreConfig(clientID: <clientID>, environment: .sandbox)

// 2. Init CardClient

let cardClient = CardClient(config: config)

// 3. Approve a card order

// Create a card from buyer's inputs
let card = Card(
    number: <cardNumber>,
    expiry: CardExpiry{
        month: <expiryMonth>,
        year: <expiryYear>
    ),
    securityCode: <securityCode>
)

// Approve the order with above card
cardClient.approveOrder(orderID: <orderID>, card: card) { result in
    switch result {
        case .success(let orderData):
            print(orderData.orderID)
            print(orderData.status)
        case .failure(let error):
            print(error)
    }
}
```

### Summary of changes

- Added:
    - `CoreConfig`
    - `CardClient` (with `approveOrder` function)
    - `Card`, `CardExpiry`, `OrderData`, `OrderStatus`
    - `CoreError` (currently having `encodingData` for cases when we fail to encode data to make a network request)
    - `ConfirmPaymentSourceRequest`, `ConfirmPaymentSourceResponse`

### Checklist

- [ ] Added a changelog entry

### Authors

- @jonathajones 
- @minhthenguyen

### CODE REVIEW

File: Sources/PaymentsCore/Models/OrderData.swift [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import Foundation
+
+public struct OrderData {
+    let orderID: String
+    let status: OrderStatus?
```
Comment (minhthenguyen) [inline Sources/PaymentsCore/Models/OrderData.swift]: currently `OrderData` only has orderID and status, but the actual network response for `/confirm-payment-source` also includes `payment-source`:
```
{
    "id": "6NV18671YP407052Y",
    "status": "APPROVED",
    "payment_source": {
        "card": {
            "last_digits": "7321",
            "brand": "VISA",
            "type": "CREDIT"
        }
    }
}
```
Should this be addressed and included in this PR?
Or leave as is for simplicity's sake and we'll consider adding this when more payment methods are implemented?

Comment (sshropshire) [reply Sources/PaymentsCore/Models/OrderData.swift]: We came across this too. I believe we flattened the response into a Kotlin `data class` in a `CardResponse` object to decouple ourselves from the RESTful response's JSON schema.

File: Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import Foundation
+
+/// Describes request to confirm a payment source (approve an order)
+public struct ConfirmPaymentSourceRequest: APIRequest {
```
Comment (sshropshire) [inline Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift]: qq: Should this be visible to the merchant?

Comment (minhthenguyen) [reply Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift]: good call, I agree this shouldn't be public!!!
But we had to since this request lives in `PaymentsCore`, but `Card` module (and others in the future) also need it :(

Comment (sshropshire) [reply Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift]: Ah yeah it'd be cool if there was a way to have it live in `Card` but sometimes the modules line up that way.

Comment (scannillo) [reply Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift]: Why does this need to live in Core? It should only be relevant to Card integrations.

File: Sources/Card/CardClient.swift [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    /// Initialize a CardClient to process card payment
+    /// - Parameter config: The CoreConfig object
+    public init(config: CoreConfig) {
+        self.config = config
+        self.apiClient = APIClient(environment: config.environment)
+    }
+
+    /// For internal use for testing/mocking purpose
+    init(config: CoreConfig, apiClient: APIClient) {
```
Comment (scannillo) [inline Sources/Card/CardClient.swift:19]:
```suggestion
    internal init(config: CoreConfig, apiClient: APIClient) {
```

File: Sources/Card/CardClient.swift [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import Foundation
+import PaymentsCore
+
+public class CardClient {
+
+    let apiClient: APIClient
+    let config: CoreConfig
```
Comment (scannillo) [inline Sources/Card/CardClient.swift]:
```suggestion
    private let apiClient: APIClient
    private let config: CoreConfig
```

File: Sources/Card/Models/CardExpiry.swift [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import Foundation
+
+public struct CardExpiry: Codable {
```
Comment (scannillo) [inline Sources/Card/Models/CardExpiry.swift]:
```suggestion
public struct ExpirationDate: Codable {
```

We can remove the `Card` prefix since it lives in the Card module. Also `ExpirationDate` is more standard, at least in the docs I've seen for BT and PP.

File: Sources/Card/CardClient.swift [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        self.apiClient = apiClient
+    }
+
+    /// Approve an order with a card, which validates buyer's card, and if valid, attaches the card as the payment source to the order.
+    /// After the order has been successfully approved, you will need to handle capturing/authorizing the order in your server.
+    /// - Parameters:
+    ///   - orderID: The ID of the order to be approved
+    ///   - card: The card to be charged for this order
+    ///   - completion: Completion handler for approveOrder, which contains data of the order if success, or an error if failure
+    public func approveOrder(orderID: String, card: Card, completion: @escaping (Result<OrderData, Error>) -> Void) {
```
Comment (scannillo) [inline Sources/Card/CardClient.swift:30]: 👏  👏  👏

File: Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift (hunk 2 above)
Comment (minhthenguyen) [reply Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift]: since we can do `confirm-payment-source` with payment sources other than card, I think this should live in `Core` so other payment methods we add in the future can use it :smile:

File: Sources/Card/CardClient.swift (hunk 3 above)
Comment (minhthenguyen) [reply Sources/Card/CardClient.swift:19]: I believe `internal` is by default

**Comment by minhthenguyen** [comment]
@scannillo @jaxdesmarais Can you help me with the CI failures? I can't seem to figure out why `Build/CocoaPods` + `Build/SPM` failed? :(

File: Sources/Card/CardClient.swift (hunk 3 above)
Comment (jaxdesmarais) [reply Sources/Card/CardClient.swift:19]: Swift defaults to internal if you don't specify an access control level. I think typically I'd lean toward not specifying internal but if we feel like it's necessary and provides clarity, happy to be overruled.

File: Sources/Card/CardClient.swift [hunk 7]
```diff
@@ -0,0 +1,52 @@
+import Foundation
+import PaymentsCore
+
+public class CardClient {
```
Comment (jaxdesmarais) [inline Sources/Card/CardClient.swift:6]: General question - does this class need to be public or can it be gated to only the Card module?

File: Sources/Card/Models/ExpirationDate.swift [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import Foundation
+
+public struct ExpirationDate: Codable {
+
+    var expiryString: String
```
Comment (jaxdesmarais) [inline Sources/Card/Models/ExpirationDate.swift]: `expiryString` > `expirationDateString`

File: Sources/Card/Models/ExpirationDate.swift [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+    /// Initialize a card expiry object
+    /// - Parameters:
+    ///   - month: The card expiration month (1 to 12)
+    ///   - year: The card expiration year in 4-digit format, `YYYY`
+    public init(month: Int, year: Int) {
+        let isSingleDigit = (1...9).contains(month)
+        let monthString = isSingleDigit ? "0\(month)" : "\(month)"
+        expiryString = "\(year)-\(monthString)"
```
Comment (jaxdesmarais) [inline Sources/Card/Models/ExpirationDate.swift]: It [looks like Android](https://github.com/paypal/Android-SDK/blob/8b86a133ff5aa2cc3027da8d8ba8c70bdffd8e61/Demo/src/main/java/com/paypal/android/ui/card/validation/DateFormatter.kt) used `/` between the month and year and formats as `month/year`. Cool if we update to match?

File: Sources/Card/CardClient.swift (hunk 3 above)
Comment (jaxdesmarais) [reply Sources/Card/CardClient.swift:19]: Just saw Minh made the same comment 4 hours ago 🤦‍♂️  probably should have refreshed the PR

File: Sources/Card/CardClient.swift (hunk 7 above)
Comment (minhthenguyen) [reply Sources/Card/CardClient.swift:6]: ah this needs to be public so merchants can use it :smile:

File: Sources/Card/Models/ExpirationDate.swift (hunk 9 above)
Comment (minhthenguyen) [reply Sources/Card/Models/ExpirationDate.swift]: I think that was for their demo app. In here we'd need to format it to `YYYY-MM` as expected by our server

File: Sources/Card/CardClient.swift (hunk 7 above)
Comment (sshropshire) [reply Sources/Card/CardClient.swift:6]: Yeah I think for the most part `CardClient` and anything in its public interface ended up needing to be `public` for Android as well. Everything else we were able to make internal, with the exception of a few classes in our `Core` module.

File: Sources/Card/Models/ExpirationDate.swift (hunk 9 above)
Comment (jaxdesmarais) [reply Sources/Card/Models/ExpirationDate.swift]: Ah, you're right, my bad! That makes sense, sorry missed that that was a UI component.

File: Sources/Card/Models/ExpirationDate.swift [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import Foundation
+
+public struct ExpirationDate: Codable {
+
+    var expirationDateString: String
```
Comment (sshropshire) [inline Sources/Card/Models/ExpirationDate.swift]: We could technically make this class internal and make `expirationDateString` available to merchants via a single property on `Card`. This would improve the onboarding experience a good bit since they'll probably make this assignment using a`TextField`'s `text` value in most cases.

File: Sources/PaymentsCore/Models/OrderData.swift [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+public struct OrderData {
```
Comment (sshropshire) [inline Sources/PaymentsCore/Models/OrderData.swift:5]: This is something Android can align on. Having a more generic `OrderData` class does make sense. On the other hand, there is some additional information from the `confirm-payment-source` endpoint response like `lastDigits` and `cardBrand` do we feel like the merchant will need this info?

File: Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift (hunk 2 above)
Comment (scannillo) [reply Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift]: 🤔  At this point, it is only being used for card checkouts - so it should live in the Card Module. We want to avoid the PaymentsCore being a catch-all for everything and utilize / lean into the modularity we have. If we come across other payment methods needing `confirm-payment-source` we can cross that bridge later. How's that sound?

File: Sources/Card/Models/Card.swift [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+public struct Card: Codable {
+
+    enum CodingKeys: String, CodingKey {
+        case number
+        case expirationDate = "expiry"
+        case securityCode = "security_code"
```
Comment (scannillo) [inline Sources/Card/Models/Card.swift]: Do we need to do all of this? Can we just do `JSONDecoder.KeyDecodingStrategy.convertFromSnakeCase`? I think we can assume things will be snake case from `orders/v2` for now. If that changes, we can look at workarounds if needed.

This would simplify our enum to then just be like this. We also get the init by default.

```
struct Card: Codable {
    case number
    case expirationDate
    case securityCode
}
```

It would also simplify things for PaymentsCore.

File: Sources/PaymentsCore/Models/OrderData.swift (hunk 11 above)
Comment (scannillo) [reply Sources/PaymentsCore/Models/OrderData.swift:5]: I think after some discussions with David we landed on a parent class living in Core, and then subclass result types living in each payment module. `CardResult`, `PayPalResult` etch to contain the more specific info.

Not sure about `OrderData` naming

File: Sources/PaymentsCore/Networking/Models/Responses/ConfirmPaymentSourceResponse.swift [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+public struct ConfirmPaymentSourceResponse: Codable {
```
Comment (scannillo) [inline Sources/PaymentsCore/Networking/Models/Responses/ConfirmPaymentSourceResponse.swift]: Can move into Card module and make private

File: Tests/UnitTests/CardTests/CardClient_Tests.swift [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    lazy var cardClient: CardClient = {
+        let config = CoreConfig(clientID: "", environment: .sandbox)
+        let apiClient = APIClient(
+            urlSession: URLSession(urlProtocol: URLProtocolMock.self),
+            environment: .sandbox
+        )
+        return CardClient(config: config, apiClient: apiClient)
+    }()
+
+    func testApproveCardOrder_withCardSuccessMockResponse_returnsValidOrderData() throws {
```
Comment (scannillo) [inline Tests/UnitTests/CardTests/CardClient_Tests.swift:16]: 👏

File: Tests/UnitTests/CardTests/CardClient_Tests.swift [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+        """
+
+        let card = Card(number: "", expirationDate: ExpirationDate(month: 1, year: 2021), securityCode: "")
+        try setupConfirmCardPaymentSourceMock(card: card, mockResponse: mockFailureResponse, statusCode: 404)
+
+        cardClient.approveOrder(orderID: "", card: card) { result in
+            guard case .failure(NetworkingError.unknown) = result else {
+                XCTFail("Expect NetworkingError.unknown for status code 404")
+                return
+            }
```
Comment (scannillo) [inline Tests/UnitTests/CardTests/CardClient_Tests.swift:74]: Should we add a TODO for adding error handling granularity for this case?

File: Tests/UnitTests/CardTests/CardClient_Tests.swift [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+            expect.fulfill()
+        }
+
+        waitForExpectations(timeout: 1)
+    }
+
+    func testApproveCardOrder_withInvalidFormatMockResponse_returnsError() throws {
```
Comment (scannillo) [inline Tests/UnitTests/CardTests/CardClient_Tests.swift]:
```suggestion
    func testApproveCardOrder_withInvalidFormatMockResponse_returnsParseError() throws {
```

File: Tests/UnitTests/PaymentsCoreTests/APIClientTests.swift [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+class APIClientTests: XCTestCase {
+
+    lazy var apiClient: APIClient = {
+        APIClient(
+            urlSession: URLSession(urlProtocol: URLProtocolMock.self),
+            environment: .sandbox
+        )
+    }()
+
+    func testAccessTokenMockRequest_Success() {
```
Comment (scannillo) [inline Tests/UnitTests/PaymentsCoreTests/APIClientTests.swift]: Naming nitpick - the SUT here is `apiClient.fetch` method

File: UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift [hunk 18]
```diff
@@ -0,0 +1,78 @@
+import XCTest
+@testable import PaymentsCore
+
+class APIClientTests: XCTestCase {
```
Comment (scannillo) [inline UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift:4]: We already have `APIClient_Tests` here (https://github.com/paypal/iOS-SDK/blob/main/Tests/UnitTests/PaymentsCoreTests/NetworkingTests/APIClient_Tests.swift). I don't think we need this additional file.

File: Tests/UnitTests/PaymentsCoreTests/APIClientTests.swift [hunk 19]
```diff
@@ -0,0 +1,56 @@
+import XCTest
+@testable import PaymentsCore
+@testable import Card
+
+class APIClientTests: XCTestCase {
```
Comment (scannillo) [inline Tests/UnitTests/PaymentsCoreTests/APIClientTests.swift]: We already have `APIClient_Tests` here (https://github.com/paypal/iOS-SDK/blob/main/Tests/UnitTests/PaymentsCoreTests/NetworkingTests/APIClient_Tests.swift). I don't think we need this additional file.

File: UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift (hunk 18 above)
Comment (scannillo) [inline UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift]: 🤔  We now have 3 files for APIClientTests at different nest levels / with different names. We should clear this up.

File: Sources/PaymentsCore/Networking/APIClient.swift [hunk 20]
```diff
@@ -24,55 +24,48 @@ public final class APIClient {
         }
 
         let task = urlSession.dataTask(with: request) { data, response, error in
-            let finish: (Result<T.ResponseType, NetworkingError>) -> Void = { result in
```
Comment (scannillo) [inline Sources/PaymentsCore/Networking/APIClient.swift:24]: Did some cleanup here to make this easier to read. 

_If this is too much/confusing for this PR, we can make a separate PR._

File: Sources/Card/Models/Card.swift (hunk 12 above)
Comment (scannillo) [reply Sources/Card/Models/Card.swift]: Steven & I updated to simplify this.

File: Sources/PaymentsCore/Networking/Models/Responses/ConfirmPaymentSourceResponse.swift (hunk 13 above)
Comment (scannillo) [reply Sources/PaymentsCore/Networking/Models/Responses/ConfirmPaymentSourceResponse.swift]: Moved into Card module.

File: Sources/Card/Models/Card.swift (hunk 12 above)
Comment (minhthenguyen) [reply Sources/Card/Models/Card.swift]: thank you :smile: !!!!

File: UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift (hunk 18 above)
Comment (scannillo) [reply UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift]: The testing structure got a little wonky after the networking PR. Also note, the file structure in XCode doesn't always accurately reflect the file structure in the repo, sometimes it gets out of sync and needs to be manually fixed.

Screen Shot 2021-10-05 at 9 55 05 AM

File: Sources/PaymentsCore/Networking/APIClient.swift (hunk 20 above)
Comment (minhthenguyen) [reply Sources/PaymentsCore/Networking/APIClient.swift:24]: thanks Sammy!! I don't think it's too much.
Although probably I'd move `correlationID` to the top since no matter whether it's error or not, as long as we have it, we should pass it to merchants

File: UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift (hunk 18 above)
Comment (scannillo) [reply UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift]: I think for this PR, let's just consolidate the APIClient_Tests into one place. In a follow up PR, let's fix this structure.

File: Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift (hunk 2 above)
Comment (minhthenguyen) [reply Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift]: yea that works for me!!

Comment (minhthenguyen) [reply Sources/PaymentsCore/Networking/Models/Requests/ConfirmPaymentSourceRequest.swift]: actually thinking a bit about it, I think `confirm-payment-source` makes more sense living in Core. a few reasons:
- it's basically just an API wrapper around an Orders v2 endpoint (and we know all endpoints work for all payment methods), plus I believe handling network requests should be Core's responsibility.
- Why do we want to move it to Card when we know for sure it works for other payment methods and we'll have to move it back to Core in the future?

Please lemme know your thought :smile:

File: UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift (hunk 18 above)
Comment (minhthenguyen) [reply UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift]: Ope this is everything I see.

Screen Shot 2021-10-05 at 11 18 43 AM

Must've been my bad fixing merge conflicts in project file, I'll get this cleanup now :D

Comment (scannillo) [reply UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift]: You're looking in the XCode navigator ^

Comment (minhthenguyen) [reply UnitTests/PaymentsCoreTests/NetworkingTests/APIClientTests.swift]: all `APIClientTests` file has been removed :smile:

File: Sources/Card/Models/Card.swift [hunk 21]
```diff
@@ -0,0 +1,32 @@
+import Foundation
+
+
+/// Initialize a card object
+/// - Parameters:
+///   - number: The card number
+///   - expiry: The card's expiration date
+///   - securityCode: The card's security code (CVV, CVC, CVN, CVE, or CID)
```
Comment (jaxdesmarais) [inline Sources/Card/Models/Card.swift]: Thoughts on adding something like "expected format" and the format to all of these? I assume there is a way that we need this data to be formatted to avoid errors.

Comment (scannillo) [reply Sources/Card/Models/Card.swift]: Great feedback. Will make this update!

File: Sources/Card/CardClient.swift [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+    ///   - completion: Completion handler for approveOrder, which contains data of the order if success, or an error if failure
+    public func approveOrder(orderID: String, card: Card, completion: @escaping (Result<OrderData, Error>) -> Void) {
+        do {
+            let confirmPaymentRequest = try ConfirmPaymentSourceRequest(
+                card: card,
+                orderID: orderID,
+                clientID: config.clientID
+            )
+
+            apiClient.fetch(endpoint: confirmPaymentRequest) { result, _ in
```
Comment (scannillo) [inline Sources/Card/CardClient.swift:38]: Do we need the `correlationID` in the `APIClient.fetch`? Here we are ignoring it.

**Review by scannillo — APPROVED** [review]

File: Sources/Card/CardClient.swift (hunk 22 above)
Comment (minhthenguyen) [reply Sources/Card/CardClient.swift:38]: yep we'll need it. 
I figure it will be somewhere in the `Result` we vend to merchants, either in `OrderData` or `Error` or both. And we have plan to discuss both `OrderData` and `Error` in both platforms so I left it unused here for now :smile:

**Review by jaxdesmarais — APPROVED** [review]
Thanks for addressing feedback! ✅ 🚀

File: Sources/Card/ConfirmPaymentSourceRequest.swift [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+            .authorization: "Basic \(encodedClientID)"
+        ]
+    }
+
+    /// Creates a request to attach a payment source to a specific order.
+    /// In order to use this initializer, the `paymentSource` parameter has to
+    /// contain the entire dictionary as it exists underneath the `payment_source` key.
+    ///
+    /// For more information on the expected body structure, see PayPal's internal
+    /// [API reference](https://ppaas/api/3719176155270329#apiReference)
```
Comment (sshropshire) [inline Sources/Card/ConfirmPaymentSourceRequest.swift]:
```suggestion
```

**Review by sshropshire — APPROVED** [review]
